### PR TITLE
Auto rechunk mechanisms

### LINKED ISSUE #3302: Method to make uniformly chunked Dask Arrays
In the process of applying some operations on Dask Arrays, chunk sizes can become a bit heterogeneous. However storage applications tend to prefer more homogeneous chunk sizes. While it is possible to the store a Dask Array to a storage format with different chunk sizes, this typically comes at a penalty of needing locking. Certainly one can rechunk the Dask Array manually. However it would be nice to have a method for Dask Arrays, which has some better ideas as to what sort of rechunking will come with low penalty. This method can then determine some reasonable homogeneous chunk size and rechunk the Dask Array using them.

### PR DESCRIPTION
**Pull request by martindurant**
Allow the passing of a string to `arr.rechunk()`, to automatically rechunk in some sensible method.

The first example shown here preserves the number of chunks in each axis, unless the given axis is already "regular" (in the sense required by saving to zarr). This is simple, easy to understand, but already useful.

Fixes #3302

- [x] Tests added / passed
- [x] Passes `flake8 dask`

### CODE REVIEW

File: dask/array/rechunk.py [hunk 1]
```diff
@@ -233,16 +251,27 @@ def rechunk(x, chunks, threshold=DEFAULT_THRESHOLD,
 
     x: dask array
         Array to be rechunked.
-    chunks:  int, tuple or dict
-        The new block dimensions to create. -1 indicates the full size of the
-        corresponding dimension.
```
Comment (shoyer) [inline dask/array/rechunk.py]: I don't think you meant to delete this from the existing docstring?

File: dask/array/rechunk.py [hunk 2]
```diff
@@ -209,6 +209,24 @@ def blockshape_dict_to_tuple(old_chunks, d):
...
+            l = len(ch)
+            newsize = int(np.ceil(size / l))
+            out.append(tuple([newsize] * (l - 1) + [size - newsize * (l - 1)]))
+        else:
+            # already regular
+            out.append(ch)
+    return tuple(out)
+
+
+rechunkers = {'preserve_number': rechunk_preserve_number}
```
Comment (shoyer) [inline dask/array/rechunk.py]: Do you have other ideas for possible "rechunkers"?

I'm not sure that this rechunking strategy is common or non-trivial enough that it is worth adding directly to `rechunk()`. Most of the time users know the chunks they want.

For a utility function like `to_zarr()`, I might write something even simpler, like `arr.rechunk([c[0] for c in arr.chunks])`. Or you could even write this function in one long line, with `arr.rechunk([int(np.ceil(s / len(c))) for s, c in zip(arr.shape, arr.chunks)])`.

Comment (martindurant) [reply dask/array/rechunk.py]: See ideas here: https://github.com/dask/dask/issues/3302#issuecomment-385674820

Comment (mrocklin) [reply dask/array/rechunk.py]: Any thoughts on shorter names?

Comment (martindurant) [reply dask/array/rechunk.py]: I wanted these to be fairly descriptive. They could be shorter if the description of what they do does into the docstring of `rechunk`.

**Comment by martindurant** [comment]
Updated following the merging of "auto" chunks, now with "simple" also included, which is `[c[0] for c in chunks]`. Not convinced, now, how useful these are.

Note that I think that the zarr PR is not dependent on anything that happens here.

**Pull request closed without merging**